core/trino-main/src/main/java/io/trino/operator/EmptyTableFunctionPartition.java

```diff
@@ -0,0 +1,118 @@
+    private final int passThroughSourcesCount;
+    private final Type[] passThroughTypes;
+
+    public EmptyTableFunctionPartition(TableFunctionProcessor tableFunction, int properChannelsCount, int passThroughSourcesCount, List<Type> passThroughTypes)
+    {
+        this.tableFunction = tableFunction;
+        this.properChannelsCount = properChannelsCount;
+        this.passThroughSourcesCount = passThroughSourcesCount;
+        this.passThroughTypes = passThroughTypes.toArray(new Type[] {});
+    }
+
+    @Override
+    public WorkProcessor<Page> toOutputPages()
+    {
+        return WorkProcessor.create(() -> {
+            TableFunctionProcessState state = tableFunction.process(null);
+            if (state.type() == TF_FINISHED) {
+                return WorkProcessor.ProcessState.finished();
+            }
+            if (state.type() == TF_BLOCKED) {
+                return WorkProcessor.ProcessState.blocked(toListenableFuture(state.blockedFuture()));
+            }
+            if (state.result() != null) {
+                return WorkProcessor.ProcessState.ofResult(appendNullsForPassThroughColumns(state.result()));
+            }
+            throw new TrinoException(FUNCTION_IMPLEMENTATION_ERROR, "When function got no input, it should either produce output or return TF_BLOCKED state");
+        });
+    }
+
+    private Page appendNullsForPassThroughColumns(TableFunctionResult result)
+    {
+        Page page = result.page();
+
+        if (page.getChannelCount() != properChannelsCount + passThroughSourcesCount) {
+            throw new TrinoException(
+                    FUNCTION_IMPLEMENTATION_ERROR,
+                    format(
+                            "Table function returned a page containing %s channels. Expected channel number: %s (%s proper columns, %s pass-through index columns)",
+                            page.getChannelCount(),
+                            properChannelsCount + passThroughSourcesCount,
+                            properChannelsCount,
+                            passThroughSourcesCount));
+        }
+
+        Block[] resultBlocks = new Block[properChannelsCount + passThroughTypes.length];
+
+        // proper outputs first
+        for (int channel = 0; channel < properChannelsCount; channel++) {
+            resultBlocks[channel] = page.getBlock(channel);
+        }
+
+        // validate that all pass-through indexes are null
+        for (int channel = properChannelsCount; channel < page.getChannelCount(); channel++) {
+            Block indexes = page.getBlock(channel);
+            for (int position = 0; position < indexes.getPositionCount(); position++) {
+                if (!indexes.isNull(position)) {
+                    throw new TrinoException(FUNCTION_IMPLEMENTATION_ERROR, "Index of a pass-through row must be null when no input data was processed. Actual: " + BIGINT.getLong(indexes, position));
+                }
+            }
+        }
```

Comment:
This is an expensive check to do on every input. If the purpose is to catch implementation errors, I think it's not worth paying the price. That should be documented as a requirement for the function implementation and it should be up to the author to ensure the implementation adheres to the contract, or risk non-sensical results or errors.

Reply:
OK, removed the check, and added a comment here as well as in the `TableFunctionProcessState` class, where the shape of output is explained. 

In fact, this is not a special requirement of the implementation. This is simply a consequence of the fact that a function can pass-through only the rows that it has seen.